File: dbxIntacctAgentTools_mcp/mcp-intacct/server/intacct/pagination.ts
/**
 * Pagination helpers for the Sage Intacct REST API.
 *
 * Mirrors intacct_sdk/pagination.py — yields every record across
 * paginated responses.
 */

export type PageFetcher<T> = (cursor: string | null) => Promise<{
  items: T[];
  nextCursor: string | null;
}>;

export interface PaginateOptions {
  maxPages?: number;
  maxResults?: number;
}

/**
 * Iterate every record across paginated REST responses.
 *
 * The fetcher takes a cursor (or null for the first page) and returns
 * { items, nextCursor }. Stops when nextCursor is null/empty or when
 * limits are hit.
 */
export async function* paginate<T>(
  fetchPage: PageFetcher<T>,
  opts: PaginateOptions = {},
): AsyncIterableIterator<T> {
  const maxPages = opts.maxPages ?? Number.POSITIVE_INFINITY;
  const maxResults = opts.maxResults ?? Number.POSITIVE_INFINITY;
  let cursor: string | null = null;
  let pageCount = 0;
  let resultCount = 0;

  while (true) {
    const { items, nextCursor } = await fetchPage(cursor);
    for (const item of items) {
      yield item;
      resultCount++;
      if (resultCount >= maxResults) {
        return;
      }
    }
    pageCount++;
    if (!nextCursor || pageCount >= maxPages) {
      return;
    }
    cursor = nextCursor;
  }
}

/** Drain a paginated iterator into an array, capped by maxResults. */
export async function collect<T>(iter: AsyncIterableIterator<T>, maxResults?: number): Promise<T[]> {
  const out: T[] = [];
  const cap = maxResults ?? Number.POSITIVE_INFINITY;
  for await (const item of iter) {
    out.push(item);
    if (out.length >= cap) {
      break;
    }
  }
  return out;
}

File: dbxIntacctAgentTools_sdk/src/intacct_sdk/pagination.py
"""Pagination helpers for the Sage Intacct REST API."""

from __future__ import annotations

from collections.abc import Callable, Iterator
from typing import Any


def paginate(
    fetch_page: Callable[[str | None], dict[str, Any]],
    *,
    items_key: str = "ia::result",
    next_cursor_key: str = "next_cursor",
    max_pages: int | None = None,
) -> Iterator[dict[str, Any]]:
    """Iterate every record across paginated REST responses.

    ``fetch_page`` is a function that takes a cursor (or None for the first
    page) and returns the parsed JSON body. ``items_key`` and
    ``next_cursor_key`` are configurable to match different REST surfaces;
    the defaults track the documented Sage Intacct shape.
    """
    cursor: str | None = None
    page_count = 0
    while True:
        body = fetch_page(cursor)
        items = body.get(items_key) or []
        for item in items:
            yield item
        cursor = body.get(next_cursor_key)
        page_count += 1
        if cursor is None or cursor == "":
            return
        if max_pages is not None and page_count >= max_pages:
            return
